java-viewer/src/main/java/org/sonar/java/viewer/DotGraph.java

```diff
@@ -25,40 +25,44 @@
 import javax.json.JsonValue;
 
 import java.text.MessageFormat;
+import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
 
-public abstract class DotDataProvider {
-
-  public static final String ESCAPE_CHAR = "?";
-
-  public abstract String label();
-
-  @CheckForNull
-  public abstract Highlighting highlighting();
-
-  @CheckForNull
-  public abstract JsonObject details();
-
-
-  public static String escape(@Nullable JsonValue jsonValue) {
-    if (jsonValue == null) {
-      return null;
+public abstract class DotGraph {
+
+  protected final List<DotGraph.Node> nodes = new ArrayList<>();
+  protected final List<DotGraph.Edge> edges = new ArrayList<>();
+
+  /**
+   * Provide the graph name
+   * @return the graph name
+   */
+  public abstract String name();
+
+  /**
+   * Fill nodes and edges for the graph
+   * @return the completed graph
+   */
+  public abstract DotGraph build();
+
+  /**
+   * Convert the CFG to DOT format (graph description language).
+   * See language specification: http://www.graphviz.org/content/dot-language
+   */
+  public final String toDot() {
+    build();
+
+    StringBuilder sb = new StringBuilder("graph " + name() + "{");
+    for (DotGraph.Node node : nodes) {
```

Comment:
matter of taste I would go for stream api here nodes.map(Node::toDot).forEach(sb::append) ? so you'll end up with two very similar lines for node and edges.